### DISCUSSION TITLE
Unable to mass install packages

### DISCUSSION THREAD

**Discussion by AlexWayfer** [pulsar-edit/.github#123] [Pulsar Support]
Hello.

I tried Pulsar. Works good, thank you.

I tried to sync my Atom settings (including customization and packages) via `sync-settings` (I'd attach a link, but I got the "Too many requests error") and got a lot of packages non-installed:

![image](https://user-images.githubusercontent.com/6510020/216323658-6be31b23-d71f-4bf8-b89f-c732132b9d12.png)

I thought they were missing in your packages repo. https://web.pulsar-edit.dev/packages does not have search. But checking by name-in-URL shows they exist.

While I tried to search `sync-settings` package to link it here, I've got the error:
![image](https://user-images.githubusercontent.com/6510020/216324356-ca9c322d-7bac-47e4-872a-fc3b263466d2.png)

So, I believe it's the reason.

Can you please relax your restrictions, rate limits, or do you have advises how should I do?

It's not a common case for newcomers or who wants to start over, but it can be a common case for migration from Atom.

**Comment by Daeraxa** [comment]
> I thought they were missing in your packages repo. [web.pulsar-edit.dev/packages](https://web.pulsar-edit.dev/packages) does not have search. But checking by name-in-URL shows they exist.

There is a search but it is on the main screen and not the package list - https://web.pulsar-edit.dev/

The rate limit is still something being tweaked, the main concern is being hit by a bot or something else that would cause the hosting costs to skyrocket for no reason. It should relax after I think 15 mins?

**Reply by AlexWayfer** [reply]
> There is a search but it is on the main screen and not the package list - https://web.pulsar-edit.dev/

OK, thanks. It'd be nice to be placed also somewhere there:

![image](https://user-images.githubusercontent.com/6510020/216327411-f972329a-2d6a-4f7d-b346-7d63830fdf74.png)


> The rate limit is still something being tweaked, the main concern is being hit by a bot or something else that would cause the hosting costs to skyrocket for no reason. It should relax after I think 15 mins?

Sad, but I got it. Thanks.

Maybe later we can use something like Cloudflare or Fastly.

**Reply by AlexWayfer** [reply]
Also, discussing packages search, they're not indexed by Google. :pensive: 

![image](https://user-images.githubusercontent.com/6510020/216333647-94498150-d1f2-4232-822c-5d5775929aa5.png)

**Reply by confused-Techie** [reply]
To echo like previously said, we are trying to find a balance for our current RateLimits, but thanks for reporting that your limit has been hit, I'll look into the request log and bump as may be needed.

Otherwise for the packages search page it may be a good idea to just include the search bar onto every single page if it's causing some confusion.

Then for your last point, are you referring to the packages themselves not being indexed by Google? Or the URL `web.pulsar-edit.dev` since it should be, I've got the Google Admin Dashboard setup for every subdomain for Pulsar so it may just be showing you a generic web result on that first option.

**Reply by AlexWayfer** [reply]
> Then for your last point, are you referring to the packages themselves not being indexed by Google?

Yep, it could help, like "pulsar linter package" query, you know.

**Reply by confused-Techie** [reply]
So for the last point I could look into getting it index, thing is we don't block it at all, so hard to say why Google hasn't just done this automatically. But something I had previously looked at was getting the search bar to appear in the search results, like you might see for other websites, that allows you to search the website directly from your Google Search Results.

But we can see what can be done

**Comment by AlexWayfer** [comment]
The limits are weird.

I've just launched Pulsar today, I think after 24 hours, and already unable to search for packages:

![image](https://user-images.githubusercontent.com/6510020/216716328-d925d1bd-856e-481a-8a85-0b9e7955fd3b.png)

**Reply by confused-Techie** [reply]
I wouldn't quite say the limits themselves are weird but Pulsar itself can be aggressive on the API.

For a quick breakdown of the API Rate Limiting

* For requests that require Authentication an IP address is allowed 300 requests within 15 minutes
* For requests that do NOT require Authentication an IP address is allowed 300 requests within 15 minutes

These two limits are individual and do not contribute to each other. 

Part of why I place the blame on Pulsar is (depending on what version you are using) when you launch Pulsar it will hit the API checking for an update to every single package. 

Then you launch the install packages page, and it hits the API for featured packages, then hits it for every package listed on that featured page (for some reason) 

Then you go to search and every single character hits the API, and if you click to install it hits for the package details, then hits again for the version data, then hits a final time for the download.

So it hits the API many times, where most of the time it seems the data it wants should already be available.

But if you are hitting it so frequently while doing just normal usage we can look at bumping it again, since we are still wanting to find the right number

**Reply by AlexWayfer** [reply]
> For a quick breakdown of the API Rate Limiting
> 
> * For requests that require Authentication an IP address is allowed 300 requests within 15 minutes
> * For requests that do NOT require Authentication an IP address is allowed 300 requests within 15 minutes
> 
> These two limits are individual and do not contribute to each other.

Seems the same. :thinking: 



> Part of why I place the blame on Pulsar is (depending on what version you are using) when you launch Pulsar it will hit the API checking for an update to every single package.

I'm using [Pulsar from the AUR (for Arch Linux), version 1.101.beta-2](https://aur.archlinux.org/packages/pulsar-bin). Has it been changed?



> But if you are hitting it so frequently while doing just normal usage we can look at bumping it again, since we are still wanting to find the right number

It was very usual usage today. If the editor checks updates via a request-per-package — it's awful and it really can hits (I think I have around 100 packages installed). But I just launched the editor, uninstalled the package and tried to search for an alternative…

**Reply by confused-Techie** [reply]
I'm not entirely sure what you mean with seems the same above.

But while we don't maintain the version on AUR the version listed there does match with the beta that has some changes to decrease the amount of hits to the API as well as caches the information received longer, so that's good.

But yeah with 100 packages it can hit that limit rather quickly. And yeah I'm not placing any blame on you for hitting it, just seems we need to bump it again then since that is very normal usage, and what we would expect.

But you are right, that it's usage of the API isn't great, we are working to fix and change as much as we can, but with so many other issues on other fronts it hasn't seen as much activity just yet.

**Reply by confused-Techie** [reply]
FYI I am currently bumping the limit for unauthenticated requests to 600 per 15 minutes per IP

**Reply by AlexWayfer** [reply]
> I'm not entirely sure what you mean with seems the same above.

I mean, the numbers ("300 requests within 15 minutes"), so I see no meaning in their differentiation, but it's your business.



> But while we don't maintain the version on AUR the version listed there does match with the beta that has some changes to decrease the amount of hits to the API as well as caches the information received longer, so that's good.

OK, I'll try to contact the AUR package maintainer.



> But yeah with 100 packages it can hit that limit rather quickly.

Got it. Sadly. For me it's pretty common usage.



> And yeah I'm not placing any blame on you for hitting it, just seems we need to bump it again then since that is very normal usage, and what we would expect.

No offenses for me, thank you for detailed quick answers. I'm trying to do my best in describing issues what I faced.

**Reply by AlexWayfer** [reply]
> > But while we don't maintain the version on AUR the version listed there does match with the beta that has some changes to decrease the amount of hits to the API as well as caches the information received longer, so that's good.
> 
> OK, I'll try to contact the AUR package maintainer.

[v1.101.0-beta](https://github.com/pulsar-edit/pulsar/releases/tag/v1.101.0-beta) is the latest official tag as I see, so I'll not disturb AUR package maintainer until there will be a new version tag.

![image](https://user-images.githubusercontent.com/6510020/216734583-d9b09a21-ed85-489f-9caf-36d8ca097e78.png)

**Reply by confused-Techie** [reply]
Ah I see what you mean, well that is to say that any endpoints that require authentication are under different RateLimits.

Such as when you publish, delete, or update a package, or log into your account. That is still rate Limited to 300 requests per 15 minutes.

Meanwhile the generic API RateLimit, such as searching for packages, getting packages, or everything else really is now Rate Limited at 600 requests per 15 minutes.

But I do see how it can be a bit confusing.

But yeah for now the tag you have seems to match our latest so no need to contact them yet. Hope things work out better from here on out for you, feel free to contact us or me again if common usage still gets you RateLimited

**Reply by AlexWayfer** [reply]
> But yeah for now the tag you have seems to match our latest so no need to contact them yet. Hope things work out better from here on out for you, feel free to contact us or me again if common usage still gets you RateLimited

You're great, thanks. I will in any (serious) case. Thank you for your work. I'll try to help as well as I can.

A little off-topic: I'm surprised about how Pulsar faster than the latest Atom (I think any engine update, like Electron, can affect this) and how it's compatible with existing packages. Very impressing, thank you all.

**Reply by Daeraxa** [reply]
The electron update did actually bump the performance a little and when we clear the path to Electron 19 it promises to actually be quite a bit faster in theory.
